### ISSUE TITLE
spec: high-availability stateless clusterized broadcasting

### ISSUE THREAD

**Issue by iameli** [livepeer/go-livepeer#1240]
## Abstract
We should allow broadcasters to maintain most of their state externally, e.g. in a database or object store.

## Motivation
Livepeer is working on our first scaled production deployment; a cluster of 15 GPU miners in a Kubernetes cluster in Chicago. This cluster will be primarily used to back transcoding plugins for Wowza and MistServer, which both utilize the broadcaster HTTP ingress at `/live` to interface with go-livepeer. We need to design our infrastructure to maximize reliability, throughput, and ease of integration.

Consistent with best practices[1] for maintaining HA services in such a context, we should run several broadcasters behind a load balancing mechanism, such that incoming requests continue to be handled even in the face of software deploys, hardware failures, and changing network conditions. One downside of operating this way is that each HTTP request could potentially hit a different broadcaster. This results in a few problems with our current broadcaster infrastructure:

1. Segments must be downloaded back from the specific broadcaster to which they were uploaded; you can't upload to one broadcaster and download from another
2. HLS manifests only exist on a per-broadcaster basis, so HLS manifests become unusable 
3. Different broadcasters will pick different orchestrators for a given stream. The resulting stream will contain visual artifacts when switching from one broadcaster to another.

1 is a blocking issue for using this architecture at all, but can be mitigated with the use of an object store. 2 is a blocker for LivepeerWowza and other HLS clients. 3 doesn't block anything but reduces the quality of the output.

## Proposed Solution

We should allow the pertinent pieces of the broadcaster's internal state to live in an external database, thus creating a "stateless broadcaster". Things this database might include:

1. URLs to segments — I'm imagining that all broadcasters using this proposal would also use an OS, so we wouldn't need to store segments themselves.
1. Orchestrator assignments, so that incoming HTTP traffic can be routed to random broadcasters and still hit the same orchestrators.

## Implementation Tasks and Considerations

I mostly don't know what to put here, but something like
1. Pick a database (Redis?)
2. Store segment information in this database
3. Store orchestrator assignments in this database so
4. Serve HLS manifests using the information in this database

## Testing Tasks and Considerations

Some integration testing, perhaps?

## Known Unknowns

I myself have no idea how complex this would be to implement in go-livepeer.

## Alternatives

1. I threw together a prototype for approximating this behavior and bootstrapping the cluster using our existing S3 store functionality, Minio, and [a lot of API proxy code](https://github.com/livepeer/livepeerjs/pull/514) to glue everything together. Maybe we shouldn't do this in go-livepeer and it should just stay in the API forever.

2. Do something other than the stateless load balancer architecture that I described.

[1]: I dunno that there's a particular canonical source for this sort of thing, but I learned a lot from https://12factor.net/

**Comment by iameli** [comment]
Would love to get @ericxtang and @j0sh's takes on this. We don't have to do all of this right away, but this is more or less my current advocacy for how we should set up broadcaster infrastructure specifically.

Also @mk-livepeer, this is somewhat similar to how you discussed setting up hosted broadcaster infrastructure when it was originally discussed for the Wowza backend months ago — main difference being that this doesn't utilize a queue. That'd be a way to further increase reliability though.

**Comment by j0sh** [comment]
> Best practices[1] for maintaining HA services in such a context would have us run several broadcasters behind a load balancing mechanism

To clarify: the goal here is to ensure data availability during broadcaster restarts and crashes? And we are *not* trying to make broadcasting workable behind a traditional load balancer? Because I agree that using a traditional LB is inadequate here, for many of the reasons you described. Moreover, load balancing via server-side proxy is neither sufficient nor necessary for HA, although that may be one component of an overall HA strategy.

So I want to drill down a bit more into *why* this strategy of using an external store for internal state is being proposed.

> HLS manifests only exist on a per-broadcaster basis, so HLS manifests become unusable

This seems to be the root of the problem. For the typical transcoding request-repsonse workflow used by Wowza and Mist Server, HLS *should not* be in the picture at all. Much of the complexity here arises from the inappropriate use of HLS in the Wowza flow. Accommodating HLS has resulted in numerous additional infrastructural moving parts: the livepeerjs API wrapper, an external object store, and (if I'm reading this proposal correctly) another external component to act as the manifest DB. This greatly increases deployment complexity, operational surface area of the entire system, injects more points of failure into the system, makes the overall data flow harder to trace through (and debug!), etc.

> This database would stuff like: Manifest URLs

Can you elaborate on what "Manifest URLs" are? Is this the contents of the manifest itself, or the addresses of segments within the manifests, or URLs to the manifests that live somewhere else?

> Store segment information in this database

What type of segment information? Metadata or the segment blobs themselves?

> Store orchestrator assignments in this database so

Can you elaborate on why we'd need to persist orchestrator assignments, especially in a closed environment such as AC?

> Alternatives
> Do something other than the stateless load balancer architecture that I described.

Rather than using an external store for internal state, another way to approach the problem is to make internal state unnecessary for HA. If achievable, this would mitigate our current concerns around data availability, without needing to introduce additional infra moving parts.

The approach to use here is traditional failover: have the client retry with another broadcaster, given an initial list of broadcasters. The client should be doing this anyway! Using an external DB / OS for internal state doesn't absolve the client from needing to retry if the broadcaster goes down mid-segment.

The use of the HLS manifest (and corresponding external DB) is not necessary when the requester can download the segment directly after receiving a response. An external object store only helps in the marginal situation where the broadcaster goes down in the short period after sending a response for a successful transcode, and before the client downloads a transcoded segment. Broadcaster reboots or crashes should be uncommon enough that the occassional slight delay from a client-side retry should adequately address these situations without additional infrastructural complexity.

That being said, there are still use cases where an external manifest or object store may still be useful [1][2][3]. But I don't think high availability is really one of those cases.

[1] https://github.com/livepeer/go-livepeer/issues/869
[2] https://github.com/livepeer/go-livepeer/issues/867
[3] https://github.com/livepeer/go-livepeer/issues/559

**Comment by iameli** [comment]
> And we are not trying to make broadcasting workable behind a traditional load balancer? Because I agree that using a traditional LB is inadequate here, for many of the reasons you described.

Ah, no, I should make that clearer. I _am_ trying to make broadcasting workable behind a traditional load balancer. High availability during restarts and crashes is one reason, I list a couple others in the proposal, but aside from those it's actually a necessity: we have no mechanism other then [metal-lb](https://metallb.universe.tf/), controlled via Kubernetes' LoadBalancer service, for external IP ingress in the Chicago cluster. We can't open ports on the servers themselves, nor (IMO) should we.

So the "reasons [I] described" are actually reasons I'm advocating for this feature, in addition to the HLS thing.

> This greatly increases deployment complexity, operational surface area of the entire system, injects more points of failure into the system, makes the overall data flow harder to trace through (and debug!), etc.

Agreed entirely, which is why I wrote this proposal to make the changes to go-livepeer directly such that the other parts can be removed. Same goal!

> This seems to be the root of the problem. For the typical transcoding request-repsonse workflow used by Wowza and Mist Server, HLS should not be in the picture at all.

This sounds like talk that's going to result in my reimplementing LivepeerWowza on every media server on the planet 😄 I get that we could make it work in this particular case, because we're the client — that's not going to be true of every single user of the system. And I still think a lot of folks are just going to point a CDN at our origin and skip things entirely.

I'd agree with you more if HLS was some wildly complex thing to support... but it's plain text output listing segments available on a local cluster of servers. That's not too bad.

> What type of segment information? Metadata or the segment blobs themselves?

Another thing I should clarify in the spec is that I figure we'd be using an OS in all cases. But I also think we should add a Redis OS, so it could end up being the same thing.

> Can you elaborate on why we'd need to persist orchestrator assignments, especially in a closed environment such as AC?

So that incoming segments can hit the load balancer, get routed to a random broadcaster, and still end up at the same orchestrator. But there are potentially other mechanisms for orchestrator persistence, this could maybe be dropped.

> The approach to use here is traditional failover: have the client retry with another broadcaster, given an initial list of broadcasters. The client should be doing this anyway! 

I don't agree; if I were implementing an API client this would feel kinda low-quality.

Other cloud services: POST the file to `https://api.example.com/file.txt`. If the upload doesn't work, retry the upload until it does.

Livepeer.live: Request a list of broadcasters from `https://livepeer.live/api/broadcaster`. POST the file to `https://chi.livepeer-ac.live/live/UUID/1.ts`. If the upload fails, request the list of broadcasters again and try the POST again.

I like the first one more. More to the point, in the context of network services, I'd call the first one "high availability" and not the second one. Load balancing lets us hit our reliability goals on a single endpoint.

> Broadcaster reboots or crashes should be uncommon enough that the occassional slight delay from a client-side retry should adequately address these situations without additional infrastructural complexity.

I'd prefer to set up infrastructure that we can deploy to completely fearlessly and run chaos monkeys and stuff. That's a good thing.

**Comment by iameli** [comment]
Rephrased a couple things to clarify intent and the data that would actually be in the database.

**Comment by j0sh** [comment]
The TL;DR here is that we really should be more precise about the failure modes we're attempting to address and the amount of fault tolerance we should expect. Zero downtime and zero data loss is probably unrealistic without sacrificing latency and throughput. Understanding this will help us refine our HA strategy and the specific approaches that are open to us.

That being said, continuing the conversation with some more pedantry:

> it's actually a necessity: we have no mechanism other then metal-lb,

This sounds more like an indictment of Kubernetes self-imposed complexity than anything else. A proxy acting as a service registrar that also preempts IP address allocation and the yak-shaving that entails? *Why?*

In any case, we should still be able to do 1-to-1 proxying (each hostname maps directly to a single B) rather than the 1-to-N as implied by "load balancing". This is preferable for a couple reasons but first I just wanted to address some of the other points around the current "difficulty" with load balancing, since I wasn't sure those were actually problems that needed solving without a LB.

I understand we're trying to hide things behind a single endpoint, but I'm not sure that's necessary at this point, or even desirable in the long term.

> Segments must be downloaded back from the specific broadcaster to which they were uploaded; you can't upload to one broadcaster and download from another

This is a strange requirement, because the client's perspective should be that the broadcaster node is the unit of work. We provide the broadcaster, and redundancies. So submitting a request to B1 but going to B2 for the results doesn't make sense. What might make more sense is the client submitting a request to B1, with the response containing a URL of where to download the results - which may be on another service such as an object store.

> Different broadcasters will pick different orchestrators for a given stream. The resulting stream will contain visual artifacts when switching from one broadcaster to another.

Visual quality is less of an issue with closed environments like AC. Orchestrator consistency is more important to amortize the overhead of starting a new GPU transcoding session.

There are ways to statelessly select an eligible set of orchestrators [1], but B consistency is important as well. The set of orchestrators a B initially receives is just a starting point: the B may be working with several Os at any point, with preferences adjusting over time based on O responsiveness, Os being added and removed from the active set based on in-flight state, errors, etc. Having to maintain all that centrally up-to-the-second is heavy and brittle. However, without that, each "new" B still needs to re-evaluate the eligible O set from scratch, which likely wouldn't yield as good results.

[1] Quick thought, randomize the initial O list based on the stream name.

> I wrote this proposal to make the changes to go-livepeer directly such that the other parts can be removed.

Not sure I understand; what parts could be removed if this proposal was implemented?

> This sounds like talk that's going to result in my reimplementing LivepeerWowza on every media server on the planet 

The original LivepeerWowza implementation was ill-advised from the beginning. Most of the issues we have there, including this proposed workaround for reliability, are a direct consequence of that initial choice to use HLS to retrieve the transcoding results. On the bright side, it's not like we have actual users depending on it right now anyway.

>  And I still think a lot of folks are just going to point a CDN at our origin and skip things entirely.

If/when that happens, we can address HLS as a *delivery* problem. HLS is mostly simple as a delivery protocol, but the semantics of its usage make it wildly inappropriate for Livepeer's transcoding protocol. LivepeerWowza issues aside, that this type of proposal is needed to make HLS delivery work reliably is a prime example that things are not as simple as they appear.

Since our current focus is on transcoding, we should address reliability within a straightforward transcoding workflow, and leave the HLS delivery for later.

> If the upload fails, request the list of broadcasters again and try the POST again.

This is really an implementation detail, but the guidance to implementers should be to retain the list of broadcasters such that clients *don't* need to request it again on each retry, because that avoids putting the JS API in the hot path during a retry. Also, it really depends on how the list itself was acquired; a hardcoded list would work just as well.

In any case, having multiple endpoints is not that unusual from a redundancy perspective. Even HLS [encourages this](https://developer.apple.com/library/archive/documentation/NetworkingInternet/Conceptual/StreamingMediaGuide/UsingHTTPLiveStreaming/UsingHTTPLiveStreaming.html#//apple_ref/doc/uid/TP40008332-CH102-SW22) for fallback playlists.

> I'd call the first one "high availability" and not the second one

That's a really narrow interpretation which discounts the remaining universe of possibilities.

We can do all kinds of things to keep a single endpoint available, but that is not as reassuring as simply providing several endpoints with minimal overlap in services. Otherwise, the user has to take it on faith that we have successfully mitigated that failure within our single endpoint by the retried request - which is not a given.

From talking with Mist, they also seem to agree with this approach, especially given the straightforward semantics of the broadcaster HTTP API: 400 error means they did something wrong, but a 500 error means Livepeer did something wrong, and a retry should be attempted.

As long as we minimize any centralized services in common between our endpoints, a retry should have higher odds of success. A database or object store is still a centralized service from the perspective of the B/O, even if the underlying implementation is distributed. (Redis is [not effective](https://aphyr.com/tags/Redis) as a distributed store, FWIW.)

> I'd prefer to set up infrastructure that we can deploy to completely fearlessly and run chaos monkeys and stuff

Sure, but if deploys and chaos testing are causing a measurable degradation in overall QoS then that's a problem. They shouldn't be. Also we really shouldn't be chaos testing a DB with a poor distributed resiliency story, so we'll likely need something else than Redis. The data store should not be exempt from failure modeling / testing.

Any choice along the axis of zero-downtime, zero-data loss will lower throughput, slow things down, and increases our overall overhead. Going all the way there is probably extreme for our purposes, so first we need to define precisely what we're willing to tolerate in terms of faults, and then we'll understand the tradeoffs we can make to achieve HA.
